File: getOptimalNum.py
def getOptimalNum(maxLength, lengthCount):
    res = []
    remainingLength = []
    for length in sorted(lengthCount, reverse=True):
        for _ in range(lengthCount[length]):
            if sum(remainingLength) == 0:
                remainingLength.append(maxLength)
                res.append([])
            hasCut = False
            for i in range(len(remainingLength)):
                if remainingLength[i] >= length:
                    remainingLength[i] -= length
                    res[i].append(length)
                    hasCut = True
                    break
            if not hasCut:
                remainingLength.append(maxLength - length)
                res.append([length])
    return res

maxLength = 3.3
lengthCount = {
    0.42: 76,
    1.294: 6,
    1.28: 2,
    1.246: 2,
    1.245: 2,
    1.402: 12,
    1.401: 2,
    1.387: 2,
    1.482: 8,
    1.342: 6,
    1.451: 10,
    1.25: 24,
}
optimalNum = getOptimalNum(maxLength, lengthCount)
totalCut = 0
totalWaste = 0
print(f"Need {len(optimalNum)} aluminium rods as follows:")
for i in range(len(optimalNum)):
    cut = sum(optimalNum[i])
    totalCut += cut
    waste = maxLength - cut
    totalWaste += waste
    print(f"{i + 1}. {str(optimalNum[i])} {cut:.2f}")
print(f"Wasted {totalWaste} in total")

File: getOptimalNum.js
function getOptimalNum(maxLength, lengthCount) {
    let res = [];
    let remainingLength = [];
    
    let lengths = Object.keys(lengthCount).map(Number).sort((a, b) => b - a);
    
    lengths.forEach(length => {
        for (let i = 0; i < lengthCount[length]; i++) {
            if (remainingLength.reduce((a, b) => a + b, 0) === 0) {
                remainingLength.push(maxLength);
                res.push([]);
            }
            let hasCut = false;
            for (let j = 0; j < remainingLength.length; j++) {
                if (remainingLength[j] >= length) {
                    remainingLength[j] -= length;
                    res[j].push(length);
                    hasCut = true;
                    break;
                }
            }
            if (!hasCut) {
                remainingLength.push(maxLength - length);
                res.push([length]);
            }
        }
    });
    
    return res;
}

function genOutput(optimalNum, maxLength) {
    let output = "";
    let totalCut = 0;
    let totalWaste = 0;
    output += `Need ${optimalNum.length} aluminium rods as follows:\n`;
    
    optimalNum.forEach((cuts, i) => {
        let cut = cuts.reduce((a, b) => a + b, 0);
        totalCut += cut;
        let waste = maxLength - cut;
        totalWaste += waste;
        output += `${i + 1}. ${JSON.stringify(cuts)} ${cut}\n`;
    });
    
    output += `Wasted ${totalWaste} in total`;
    return output;
}

document.getElementById('btn').addEventListener('click', function() {
    const maxLength = 3.3;
    const lengthCount = {
        0.42: 76,
        1.294: 6,
        1.28: 2,
        1.246: 2,
        1.245: 2,
        1.402: 12,
        1.401: 2,
        1.387: 2,
        1.482: 8,
        1.342: 6,
        1.451: 10,
        1.25: 24,
    };

    const optimalNum = getOptimalNum(maxLength, lengthCount);
    const output = genOutput(optimalNum, maxLength);
    document.getElementById('output').textContent = output;
});
